### ISSUE TITLE
Response data outside the window of the actual query

### ISSUE THREAD

**Issue by jacksontj** [Comcast/trickster#96]
I ran into an issue with trickster where I'm sending a query_range for a specific range (A-B) and I'm getting an entire response that is outside of the range. I was easily able to reproduce this locally with 2 queries. First you send a query with a plausible time range, lets say `start=1543010160&end=1543012666&step=45` this will return some results, in my example the first timestamp is `1543010130` (its a little off due to the fastforward, this is not great but not the main problem). Then if we send a query for something obviously outside the range like `start=1643010160&end=1643012666&step=45` we get an identical response, which is wrong for a variety of reasons, not the least of which being that the time ranges don't overlap (not to mention the fact that the second range is in the year 2022 ;) ).

While looking into this I found that the `deriveCacheKey` function doesn't take start/stop into account, only the `time` query param. This is an issue because query_range *only* has start/stop (not time). This seems like an odd bug to me because I would have figured we'd run into this all the time. The cache key does include the `step` parameter, so if that varies in practice then this issue would be harder to run into (which lines up with the behavior I've seen using grafana -- sometimes changing the range gives more or less data).

So I locally added the start/stop time to the cache key (https://github.com/Comcast/trickster/compare/master...jacksontj:range_cache?expand=1) but I'm not sure if this is the *correct* fix. It seems to work, but I'm a bit uneasy as this seems like such a glaringly obvious bug that I'm probably missing something.

**Comment by jranson** [comment]
Hey Thomas thanks for the report - appreciate you continuing to use Trickster and report issues back!

I will take a look at the unexpected results and see if I can figure out why and correct. I don't think the issue is in `deriveCacheKey` but rather in the date parsing/manipulation. Here's the tldr on why:

When it is present, the `time` parameter is considered in `deriveCacheKey`, since it indicates the user is requesting an instantaneous value representing some absolute point in time.  We temporarily cache instantaneous values (for 15s by default) in a separate record from the series range values, since they are short-lived and not rounded via step boundary normalization. So using the `time` parameter allows us to cache instantaneous values for the same query, for different times, concurrently, as each will have its own small cache record. We discovered that is how some users' dashboards are designed (e.g., multiple instantaneous values using the same query but with different time params) and made that adjustment.

For series range values, however, we do not (and should not) consider the start/end parameters in `deriveCacheKey` in the same way we consider the `time` param for instantaneous, as we want all data points in a given series to go into the same cache record, regardless of what time range the end user actually requests. That way, subsequent requests for the same query (having the same or different start/ends) can pull any relevant data points from the existing cache (we crop the user response resultset back down to only their requested time range before sending) and only ask prometheus for what is missing.  If we were to consider start/end, that would effectively make Trickster useless, since the derived cache record hash can no longer be predictably generated, as the timestamps, while moving chronologically, will be random and create quasi-random key name hashes.  For example, consider a Grafana dashboard while I am looking at a 6-hour view, and you have the same dashboard up on a 12-hour view. Since your start/end are different from mine, we would not be using the same cache if those params are considered when deriving the cache key. And in a few minutes when the normalized step boundary moves to the next pole, we'd both then have a full cache miss, execute the full query to prometheus, and then cache in separate records again. Not good.

In any event, I'll take a look at the behaviors of the date parsing/manipulation and see if there is anything there that would be a problem.

**Comment by jacksontj** [comment]
From another quick look, it seems that `cropToRange` doesn't remove series that have no data within the range. Specifically, if the dataset has only data outside the range it doesn't actually crop anything -- so in this case since we have *no* data in the range it doesn't crop any of the data out-- leaving us with the original (incorrect) result.

**Comment by jacksontj** [comment]
I have updated my branch to make the `cropToRange` function crop data even if nothing matches.

**Comment by jranson** [comment]
Ah! Good find. Did this solve your issue? If so, can you submit a PR? Thanks again for all your help!

**Comment by jacksontj** [comment]
I did some more looking and I found another related issue. The PR I have open fixes the case where it doesn't filter the data properly. But as it stands today if there was an *error* fetching some part of the data, but *any* data with that cachekey was returned, it returns an empty success response (instead of the error from prom). I have updated my PR to include a second commit which covers that case.
